File: EasyDataAPIWrapper.Java/src/main/java/services/ApiClient.java
package services;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.type.CollectionType;
import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import models.Crud;
import models.Method;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ApiClient {
    private String _port = "5000";
    private String _host = "localhost";
    private String _protocol = "http";
    private String _baseUrl = "api/easydata/models/__default/sources";
    private String _dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
    private String _jsonCharset = "utf-8";

    public void setPort(String _port) {
        this._port = _port;
    }

    public void setHost(String _host) {
        this._host = _host;
    }

    public void setProtocol(String _protocol) {
        this._protocol = _protocol;
    }

    public void setBaseUrl(String _baseUrl) {
        this._baseUrl = _baseUrl;
    }

    public void setDateFormat(String _dateFormat) {
        this._dateFormat = _dateFormat;
    }

    public void setJsonCharset(String _jsonCharset) {
        this._jsonCharset = _jsonCharset;
    }

    public <T> String generateURI(Class<T> clazz, Crud crud) {
        return String.format(_protocol + "://" + _host + ":" + _port + "/" + _baseUrl + "/" + clazz.getSimpleName() + "/" + crud.toString());
    }

    public <T> List<T> get(Class<T> clazz) {
        try {
            String uri = generateURI(clazz, Crud.get);

            NetworkThread t = new NetworkThread(uri, "", Method.GET.toString(), _jsonCharset);
            t.start();
            t.join();
            String jsonResponse = t.getResult();

            List<T> array = jsonArrayToList(jsonResponse, clazz);

            return array;
        } catch (Exception ex) {
            System.out.println(ex);
        }
        return null;
    }

    public <T> List<T> jsonArrayToList(String json, Class<T> elementClass) throws IOException {
        ObjectMapper objectMapper = new ObjectMapper();
        CollectionType listType =
                objectMapper.getTypeFactory().constructCollectionType(ArrayList.class, elementClass);
        return objectMapper.readValue(json, listType);
    }

    public int create(Object obj) {
        try {
            Gson gson = new GsonBuilder()
                    .setDateFormat(_dateFormat)
                    .setFieldNamingPolicy(FieldNamingPolicy.UPPER_CAMEL_CASE)
                    .create();

            String body = gson.toJson(obj);

            String uri = generateURI(obj.getClass(), Crud.create);

            NetworkThread t = new NetworkThread(uri, body, Method.POST.toString(), _jsonCharset);
            t.start();
            t.join();
            String jsonResponse = t.getResult();

            JsonObject jsonObj = gson.fromJson(jsonResponse, JsonObject.class);
            return jsonObj.get("record").getAsJsonObject().get("Id").getAsInt();
        } catch (Exception ex) {
            System.out.println(ex);
        }
        return 0;
    }

    public <T> int update(T obj) {
        try {
            Gson gson = new GsonBuilder()
                    .setDateFormat(_dateFormat)
                    .setFieldNamingPolicy(FieldNamingPolicy.UPPER_CAMEL_CASE)
                    .create();

            String body = gson.toJson(obj);

            String uri = generateURI(obj.getClass(), Crud.update);

            NetworkThread t = new NetworkThread(uri, body, Method.UPDATE.toString(), _jsonCharset);
            t.start();
            t.join();
            String jsonResponse = t.getResult();


            JsonObject jsonObj = gson.fromJson(jsonResponse, JsonObject.class);
            return jsonObj.get("record").getAsJsonObject().get("Id").getAsInt();
        } catch (Exception ex) {
            System.out.println(ex);
        }
        return 0;
    }

    public <T> void delete(int id, Class<T> clazz) {
        try {
            String uri = generateURI(clazz, Crud.delete);
            String body = "{Id: " + id + "}";

            NetworkThread t = new NetworkThread(uri, body, Method.DELETE.toString(), _jsonCharset);
            t.start();
        } catch (Exception ex) {
            System.out.println(ex);
        }
    }
}

File: EasyDataApiWrapper.C#/EasyDataApiWrapper/services/ApiClient.cs
using EasyDataApiWrapper.models;
using Newtonsoft.Json;
using System.Net;

namespace EasyDataApiWrapper.services
{
    public class ApiClient
    {

        private string _port = "5000";
        private string _host = "localhost";
        private string _protocol = "http";
        private string _baseUrl = "api/easydata/models/__default/sources";
        private DateTimeZoneHandling _dateTimeZoneHandling = DateTimeZoneHandling.Utc;
        private string _jsonCharset = "utf-8";

        public string Port
        {
            set { _port = value; }
        }

        public string Host
        {
            set { _host = value; }
        }

        public string Protocol
        {
            set { _protocol = value; }
        }

        public string BaseUrl
        {
            set { _baseUrl = value; }
        } 
        
        public string JsonCharset
        {
            set { _jsonCharset = value; }
        }

        public DateTimeZoneHandling DateTimeZoneHandling 
        {
            set { _dateTimeZoneHandling = value; }
        }

        private string GenerateURI<T>(Crud crud)
        {
            return $"{_protocol}://{_host}:{_port}/{_baseUrl}/{typeof(T).Name}/{crud}";
        }


        private string SendRequest(string? body, string uri, Method method)
        {
            try
            {
                var httpWebRequest = (HttpWebRequest) WebRequest.Create(uri);
                httpWebRequest.ContentType = $"application/json; charset={_jsonCharset}";
                httpWebRequest.Method = method.ToString();

                if (body?.Length > 0)
                {
                    using (var streamWriter = new StreamWriter(httpWebRequest.GetRequestStream()))
                    {
                        streamWriter.Write(body);
                        streamWriter.Flush();
                    }
                }


                var httpResponse = (HttpWebResponse)httpWebRequest.GetResponse();
                using (var streamReader = new StreamReader(httpResponse.GetResponseStream()))
                {
                    var result = streamReader.ReadToEnd();
                    return result;
                }
            }
            catch (Exception ee)
            {
                return ee.ToString();
            }
        }


        public List<T>? Get<T>()
        {
            try
            {
                string uri = GenerateURI<T>(Crud.get);
                string json = SendRequest(null, uri, Method.POST);
                return JsonConvert.DeserializeObject<List<T>>(json);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
            return null;
        }


        public int Create<T>(T obj)
        {
            try
            {
                string json = JsonConvert.SerializeObject(obj, new JsonSerializerSettings
                {
                    DateTimeZoneHandling = _dateTimeZoneHandling
                });

                string uri = GenerateURI<T>(Crud.create);
                string jsonRequest = SendRequest(json, uri, Method.POST);

                dynamic? jsonObj = JsonConvert.DeserializeObject(jsonRequest);
                return Convert.ToInt32(jsonObj?["record"]["Id"]);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
            return 0;
        }

        public int Update<T>(T obj)
        {
            try
            {
                string json = JsonConvert.SerializeObject(obj, new JsonSerializerSettings
                {
                    DateTimeZoneHandling = DateTimeZoneHandling.Utc
                });

                string uri = GenerateURI<T>(Crud.update);
                string jsonRequest = SendRequest(json, uri, Method.POST);

                dynamic? jsonObj = JsonConvert.DeserializeObject(jsonRequest);
                return Convert.ToInt32(jsonObj?["record"]["Id"]);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }

            return 0;
        }


        public void Delete<T>(int id)
        {
            try
            {
                string json = "{ Id: " + id + " }";

                string uri = GenerateURI<T>(Crud.delete);
                string jsonRequest = SendRequest(json, uri, Method.DELETE);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
        }
    }
}
